### PR TITLE
[Config] Add optional arguments

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR does two things:

- 1. Allows to add optional keyword arguments to pipeline configs
- 2. Make the logic of optional components easier 

**Import**: Merging this PR assumes that all keyword arguments are **not** loadable weights, components. IMO this makes sense since one would never set a unet to a default value, but please pay special attention to this when reviewing.

You can try it out by doing:

```python
#!/usr/bin/env python3
from diffusers import DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained("diffusers/tiny-stable-diffusion-torch")

pipeline.save_pretrained("hey-test")

pipeline2 = DiffusionPipeline.from_pretrained("hey-test")
```

I'll add tests as soon as you're ok with the design :-)

### CODE REVIEW

File: src/diffusers/pipeline_utils.py [hunk 1]
```diff
@@ -664,6 +677,14 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         model = pipeline_class(**init_kwargs)
         return model
 
+    @staticmethod
+    def _get_signature_keys(obj):
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:681]: this function is very important - it defines what keyword args are modules and which are not cc @patil-suraj @pcuenca @anton-l

File: src/diffusers/pipeline_utils.py [hunk 2]
```diff
@@ -129,10 +129,14 @@ class DiffusionPipeline(ConfigMixin):
...
 
-        - **config_name** ([`str`]) -- name of the config file that will store the class and module names of all
+        - **config_name** (`str`) -- name of the config file that will store the class and module names of all
           components of the diffusion pipeline.
+        - **_optional_components** (List[`str`]) -- lists of all components that are optional so which don't have to be
+          passed for the pipeline to function (should be overridden by parent class).
     """
     config_name = "model_index.json"
+    _optional_components = []
+    _pipeline_kwargs = []
```
Comment (pcuenca) [inline src/diffusers/pipeline_utils.py:139]: What is this used for?

File: src/diffusers/pipeline_utils.py [hunk 3]
```diff
@@ -184,12 +188,18 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike]):
         model_index_dict.pop("_diffusers_version")
         model_index_dict.pop("_module", None)
 
+        def is_saveable_module(name, value):
+            if not isinstance(value, (list, tuple)):
```
Comment (pcuenca) [inline src/diffusers/pipeline_utils.py:192]: What if one of the `passed_pipe_kwargs` is a list or a tuple?

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 4]
```diff
@@ -114,7 +116,7 @@ def __init__(
             new_config["clip_sample"] = False
             scheduler._internal_dict = FrozenDict(new_config)
 
-        if safety_checker is None:
+        if safety_checker is None and requires_safety_checker:
```
Comment (pcuenca) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:119]: I don't think this is robust enough, or I'm not fully understanding. Consider the following case:

- `requires_safety_checker` is `False`
- The pre-trained repo does contain a safety checker model.
- The user passes `safety_checker=None` to `from_pretrained`

Wouldn't the warning still be displayed by `DiffusionPipeline.from_pretrained`?  https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipeline_utils.py#L573

**Review by pcuenca — COMMENTED** [review]
Looks complicated to me, it's hard to keep in your mind all the code paths involved. I may not be fully understanding it, but it also appears finicky with the assumptions about the pipeline kwargs.

I'm not sure how I'd handle it differently, though. I'll think about it later. I'd appreciate any additional insights from others here :)

File: src/diffusers/pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipeline_utils.py:139]: True this is not used and can be removed

File: src/diffusers/pipeline_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/pipeline_utils.py:192]: yeah that's a bit problematic at the moment indeed

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:119]: Hmm no if `requires_safety_checker` is False then the warning cannot be displayed - can you add a repro code snippet?

File: src/diffusers/pipeline_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/pipeline_utils.py:192]: Changing this!

File: src/diffusers/pipeline_utils.py [hunk 5]
```diff
@@ -184,12 +187,20 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike]):
         model_index_dict.pop("_diffusers_version")
         model_index_dict.pop("_module", None)
 
+        expected_modules, optional_kwargs = self._get_signature_keys(self)
+
+        def is_saveable_module(name, value):
+            if name not in expected_modules:
+                # modules have to be saved in list format
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:194]:
```suggestion
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 6]
```diff
@@ -83,6 +84,7 @@ def __init__(
         ],
         safety_checker: StableDiffusionSafetyChecker,
         feature_extractor: CLIPFeatureExtractor,
+        requires_safety_checker: bool = True,
```
[the remark this answers is not in the document]
Comment (anton-l) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:87]: ok, I see that it's used for the safety error, but maybe setting the default to `None` would help to simplify some of the loading logic

[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:87]: Hmm I'd not like to mix the two to be honest. One is components, modules, one is optional kwargs (ints, list)

**Comment by patrickvonplaten** [comment]
Added some tests that should help understand the use case a bit better :-)

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 4 above)
Comment (pcuenca) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:119]:
```Python
pipeline = DiffusionPipeline.from_pretrained("diffusers/tiny-torch-full-checker", safety_checker=None)
```

It's like the tiny model but with a safety checker (sorry I uploaded a full-sized one). Also, `requires_safety_checker` is `False` in the config.

Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:119]: Your command above doesn't produce a warning as it shouldn't

**Comment by averad** [comment]
@patrickvonplaten 

```
Traceback (most recent call last):
pipe = OnnxStableDiffusionPipeline.from_pretrained(
  File "\lib\site-packages\diffusers\pipeline_utils.py", line 672, in from_pretrained
    raise ValueError(
ValueError: Pipeline <class 'diffusers.pipelines.stable_diffusion.pipeline_onnx_stable_diffusion.OnnxStableDiffusionPipeline'> expected {'feature_extractor', 'safety_checker', 'text_encoder', 'vae_encoder', 'unet', 'vae_decoder', 'tokenizer', 'scheduler'}, but only {'safety_checker', 'text_encoder', 'vae_encoder', 'unet', 'vae_decoder', 'tokenizer', 'scheduler'} were passed.
```

File: src/diffusers/pipeline_utils.py [hunk 7]
```diff
@@ -582,7 +582,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                             f"{passed_class_obj[name]} is of type: {type(passed_class_obj[name])}, but should be"
                             f" {expected_class_obj}"
                         )
-                elif passed_class_obj[name] is None:
+                elif passed_class_obj[name] is None and name not in pipeline_class._optional_components:
```
Comment (pcuenca) [inline src/diffusers/pipeline_utils.py:584]: 👍

**Pull request merged**